File: IJA/projekt/src/sample/Coordinate.java
/**
 * Coordinate - trida pro praci se souradnicemi
 * @author Michal Dostal (xdosta51)
 * @author Radek Svec (xsvecr01)
 * 
 */

package sample;

public class Coordinate{

    private double x = 0;
    private double y = 0;

    /**
     * Konstruktor pro tridu Coordinate
     * @param cx double bod souradnice na ose X
     * @param cy double bod souradnice na ose Y
     */
    protected Coordinate(double cx, double cy){
        this.x = cx;
        this.y = cy;
    }

    /**
     * 
     * @return double vraci bod souradince na ose X
     */
    public double getX(){
        return this.x;
    }

    /**
     * 
     * @return double vraci bod souradnice na ose Y
     */
    public double getY(){
        return this.y;
    }

    /**
     * Vypocita vzdalenost na ose X mezi 2 Coordinaty
     * @param c Coordinate porovnavana souradnice Coordinate
     * @return double vraci vzdalenost na ose X
     */
    public double diffX(Coordinate c){
        return this.x - c.getX();
    }

    /**
     * Vypocita vzdalenost na ose Y mezi 2 Coordinaty
     * @param c Coordinate porovnavana souradnice Coordinate
     * @return double vraci vzdalenost na ose Y
     */
    public double diffY(Coordinate c){
        return this.y - c.getY();
    }

    /**
     * Porovna dva objekty Coordinate jestli se rovnaji
     * @param obj Object porovnavany objekt Coordinate
     * @return boolean vraci true pokud se souradnice rovnaji, jinak false
     */
    public boolean equals(Object obj){
        if(obj == this) {
            return true;
        }
        if(obj == null || obj.getClass() != this.getClass()){
            return false;
        }
        Coordinate cordo = (Coordinate) obj;
        if(this.x == cordo.getX() && this.y == cordo.getY()){
            return true;
        }
        else return false;
    }

    /**
     * Override pro hashCode
     * @return int vraci vypocitany hash code
     */
    public int hashCode() {
        double result = 17;
        result = 31 * result + x;
        result = 31 * result + y;
        return (int)result;
    }

    /**
     * Funkce pro vytvoreni nove souradnice se zadanymi body X a Y
     * @param cx double bod na ose X vytvarene souradnice
     * @param cy double bod na ose Y vytvarene souradnice
     * @return Coordinate vraci vytvoreny objekt
     */
    public static Coordinate create(double cx, double cy) {
        if(cx >= 0 && cy >= 0){
            return new Coordinate(cx, cy);
        }
        else{
            return null;
        }
    }
}

File: ICP/xsvecr01-xpospe05/src/coordinate.cpp
/**
 * @file coordinate.cpp
 * @author Jan Pospech [xpospe05]
 * @author Radek Svec [xsvecr01]
 * @brief Trida souradnice. 
 * @version 0.1
 * @date 2020-05-12
 * 
 * @copyright Copyright (c) 2020
 * 
 */

#include "coordinate.h"
#include "math.h"

Coordinate::Coordinate(){
    this->x = 0;
    this->y = 0;
}

Coordinate::Coordinate(int cx, int cy){
    this->x = cx;
    this->y = cy;
}

int Coordinate::getX(){
    return this->x;
}
int Coordinate::getY(){
    return this->y;
}

int Coordinate::diffX(Coordinate c){
    return (this->x - c.getX());
}
int Coordinate::diffY(Coordinate c){
    return (this->y - c.getY());
}

float Coordinate::distance(Coordinate c){
    float result = sqrt(pow(this->diffX(c), 2) + pow(this->diffY(c), 2));
    if (result < 0) result = -result;
    return result;
}

float Coordinate::angle(Coordinate c){
    return ((atan2(this->diffY(c), this->diffX(c)) * 180) / M_PI);
}

bool Coordinate::equals(Coordinate c){
    if((this->getX() == c.getX()) && (this->getY() == c.getY())){
        return true;
    }
    else return false;
}
